Copying link to a common term in an article results in the incorrect part of the page being highlighted

hortont424 opened the pull request:
#### 42ee0a01c846807e6b64f348ffa141e05e8188d4
<pre>
Copying link to a common term in an article results in the incorrect part of the page being highlighted
<a href="https://bugs.webkit.org/show_bug.cgi?id=279248">https://bugs.webkit.org/show_bug.cgi?id=279248</a>
<a href="https://rdar.apple.com/134882107">rdar://134882107</a>

Reviewed by Richard Robinson and Megan Gardner.

The fragment directive generation code was terminating its context walk when it hit
*any* node boundary, meaning that (for example) if you selected a word in its own &lt;b&gt;,
we wouldn&apos;t be able to generate any context, and the link would match the first
instance of that word on the page, not the correct one.

The intent was to match the fragment directive *search* code, which breaks across
block boundaries, not any element&apos;s boundary. Factor out this code (and add a
place where we can share code between generation and search, since we will likely
need more of that in the future).

* LayoutTests/http/tests/scroll-to-text-fragment/generation-can-pass-inline-boundaries-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/generation-can-pass-inline-boundaries.html: Added.
Add a test.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
(WebCore::positionsHaveSameBlockAncestor):
(WebCore::previousWordsFromPositionInSameBlock):
(WebCore::nextWordsFromPositionInSameBlock):
Make the aforementioned change, only terminating the context search when we exit the block.

(WebCore::FragmentDirectiveGenerator::generateFragmentDirective):
Make sure layout is up to date before we go reading from the render tree.
Move the fragment directive prefix string closer to its use.

(WebCore::previousWordsFromPosition): Deleted.
(WebCore::nextWordsFromPosition): Deleted.

* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::nearestBlockAncestor): Deleted.
* Source/WebCore/dom/FragmentDirectiveUtilities.cpp: Added.
(WebCore::FragmentDirectiveUtilities::nearestBlockAncestor):
* Source/WebCore/dom/FragmentDirectiveUtilities.h: Added.
Factor out nearestBlockAncestor. I couldn&apos;t find anything similar in the rest
of the project, so for now kept it local to this code.

Canonical link: <a href="https://commits.webkit.org/283294@main">https://commits.webkit.org/283294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/780f44b7477b7b632a816275c8cdaf853b74073b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18510 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69920 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16500 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52875 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11460 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68958 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41775 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57040 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33510 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14417 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15376 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60286 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14763 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71623 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9846 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9878 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60479 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14518 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8129 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1764 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41072 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42148 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -49,12 +51,17 @@ FragmentDirectiveGenerator::FragmentDirectiveGenerator(const SimpleRange& textFr
-static String previousWordsFromPosition(unsigned numberOfWords, VisiblePosition& startPosition)
+static bool positionsHaveSameBlockAncestor(const VisiblePosition& a, const VisiblePosition& b)
+{
+    return nearestBlockAncestor(*a.deepEquivalent().containerNode()) == nearestBlockAncestor(*b.deepEquivalent().containerNode());
```
why is it safe to dereference these without null checks?

hortont424 replied:
fair!

rr-codes commented on Source/WebCore/dom/FragmentDirectiveUtilities.cpp:
```diff
@@ -0,0 +1,51 @@
+const Node& nearestBlockAncestor(const Node& node)
+{
+    const Node* currentNode = &node;
+    while (currentNode) {
```
I slightly prefer `for auto* currentNode = &node; currentNode; currentNode = currentNode->parentNode();` but it doesn't matter that much.

Also, should this use a smart pointer?

hortont424 replied:
I am just moving this code :) (the answers are probably both yes)

rr-codes commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -141,6 +149,7 @@ void FragmentDirectiveGenerator::generateFragmentDirective(const SimpleRange& te
     if (suffix)
         components.append(makeString('-', *suffix));
 
+    const String textDirectivePrefix = ":~:text="_s;
```
can this be a constexpr ASCIILIteral instead>

hortont424 replied:
again with the moving :) there's a lot in this code that could be much neater. but I'll see if I can make some of these changes

pxlcoder replied:
`constexpr auto` if making the change.

rr-codes replied:
will `auto` know to make an ASCIILiteral? guess we'll find out

hortont424 replied:
that is what `operator"" _s` returns so it sure should

rr-codes replied:
ah, I always mistakenly think it returns a String

megangardner commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -141,6 +151,7 @@ void FragmentDirectiveGenerator::generateFragmentDirective(const SimpleRange& te
     if (suffix)
         components.append(makeString('-', *suffix));
 
+    static constexpr auto textDirectivePrefix = ":~:text="_s;
```
Is this not defined somewhere else? I swear it was...

aprotyas replied:
This is the only place in fact!

hortont424 committed:
Copying link to a common term in an article results in the incorrect part of the page being highlighted
https://bugs.webkit.org/show_bug.cgi?id=279248
rdar://134882107

Reviewed by Richard Robinson and Megan Gardner.

The fragment directive generation code was terminating its context walk when it hit
*any* node boundary, meaning that (for example) if you selected a word in its own <b>,
we wouldn't be able to generate any context, and the link would match the first
instance of that word on the page, not the correct one.

The intent was to match the fragment directive *search* code, which breaks across
block boundaries, not any element's boundary. Factor out this code (and add a
place where we can share code between generation and search, since we will likely
need more of that in the future).

* LayoutTests/http/tests/scroll-to-text-fragment/generation-can-pass-inline-boundaries-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/generation-can-pass-inline-boundaries.html: Added.
Add a test.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
(WebCore::positionsHaveSameBlockAncestor):
(WebCore::previousWordsFromPositionInSameBlock):
(WebCore::nextWordsFromPositionInSameBlock):
Make the aforementioned change, only terminating the context search when we exit the block.

(WebCore::FragmentDirectiveGenerator::generateFragmentDirective):
Make sure layout is up to date before we go reading from the render tree.
Move the fragment directive prefix string closer to its use.

(WebCore::previousWordsFromPosition): Deleted.
(WebCore::nextWordsFromPosition): Deleted.

* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::nearestBlockAncestor): Deleted.
* Source/WebCore/dom/FragmentDirectiveUtilities.cpp: Added.
(WebCore::FragmentDirectiveUtilities::nearestBlockAncestor):
* Source/WebCore/dom/FragmentDirectiveUtilities.h: Added.
Factor out nearestBlockAncestor. I couldn't find anything similar in the rest
of the project, so for now kept it local to this code.

Canonical link: https://commits.webkit.org/283294@main
